torch/csrc/autograd/variable.cpp

```diff
@@ -151,51 +78,43 @@ void Variable::backward(
 
   std::vector<Variable> inputs;
   if (!gradient.has_value()) {
-    gradient = make_variable(at::ones_like(data()), /*requires_grad=*/false);
+    gradient = make_variable(at::ones_like(*this), /*requires_grad=*/false);
   }
   inputs.push_back(std::move(as_variable_ref(*gradient)));
   Engine::get_default_engine().execute(edges, inputs, keep_graph, create_graph);
 }
 
-void Variable::Impl::set_data(Tensor new_data) {
+void Variable::set_data(Tensor new_data) const {
+  AT_ASSERT(is_sparse() == new_data.is_sparse());
+
   // Resets gradient accumulator if metadata is out of date
-  auto autograd_meta = get_autograd_meta();
+  Variable::AutogradMeta* autograd_meta = get_autograd_meta();
   std::lock_guard<std::mutex> lock(autograd_meta->mutex_);
   auto prior_accumulator = autograd_meta->grad_accumulator_.lock();
   if (prior_accumulator) {
     const auto prior_device = prior_accumulator->input_metadata(0).device();
     const auto new_device = new_data.is_cuda() ? new_data.get_device() : -1;
 
-    if (new_data.type() != data_.type() || prior_device != new_device) {
+    if (new_data.type() != type() || prior_device != new_device) {
       autograd_meta->grad_accumulator_.reset();
     }
   }
 
-  // Updates metadata
-  data_type_ = new_data.type().typeMeta();
-  type_id_ = new_data.type().type_id();
-  is_variable_ = true;
-
-  auto new_data_copy = at::Tensor(new_data.getIntrusivePtr()->shallow_copy_and_detach());
-  data_ = std::move(new_data_copy);
-}
-
-void Variable::Impl::release_resources() {
-  autograd_meta_.reset();
-  data_.reset();
+  get()->shallow_copy_from(new_data.getIntrusivePtr()->shallow_copy_and_detach());
+  get()->set_autograd_meta(std::move(get()->detach_autograd_meta()));
+  get()->set_is_variable(true);
```

Comment:
This looks wrong. If it wasn't a variable before, it still shouldn't be a variable now.